### DISCUSSION TITLE
Are one-liner solutions the way to go?

### DISCUSSION THREAD

**Discussion by Picoquio** [codewars/codewars.com#2637] [General]
Hi there! 

I see that the most upvoted solutions tend to be one-liner solutions, or those that take as few lines as possible. Sometimes these solutions are harder to understand compared to those that are a bit lengthier. Some of these solutions even had tons of "best practices" upvotes, when they clearly shouldn't have (bitwise operators for a 6-7 kyu kata, seriously?). Every time I submit my solution I think something on the lines of "good, now lets see how others did it in less lines than me". It's always good to learn from others of course, but I'm worried I might be getting into the wrong mindset. Is less code always better? Which one to choose: readability or efficiency? 

Thank you, and sorry if this has been asked before.

**Comment by Nik-Sharp** [comment]
**Is less code always better?**

A good amount of the time, mainly on harder katas, the "short" solutions will either not be readable, have worse performance, or not have either. Of course, the easier the kata, the more likely it is that you can make a short solution that performs well and is readable, but not always

 **Which one to choose: readability or efficiency**

I would say try to strike a balance between both. But shortcode isn't always faster or easier to read, and sometimes, the longer code is faster and more readable than the extremely short solutions.

**Comment by vuurball** [comment]
readability should be top priority in real life
I've been coding for over 10 years professionally, every day, and I never code the way I answer katas here lol.. come to think of it, I rarely needed to join and split arrays as much as here :)

here are 2 main differences popping to mind:
1. in real world application your code is never 1 function, it's hundreds & thousands... imagine all of them looking like the answers here.. 
2. in real life there is no detailed explanation next to the code... so if it's hard to understand, it's bad code, even if it's considered top rated answer here. (and if your code requires 10 lines of comments to explain it, that's also bad)

good luck with your learning :)

**Comment by juvenigi** [comment]
1. "best practice" should be taken with a pinch of salt as it depends on who you ask.
2. IMO, one-liners are not bad as long as the code is *descriptive*
- example of descriptive ```if( isCool(x) )```, phrasing a function isSomething naturally leads one to believe the return type is a boolean, no need to do ```isCool(x) == true```
- example of non-descriptive but often used in fancy one-liners ```!!n``` (true if not null in javascript)
3. even if it's short it might have bad performance
- take swapping values of two elements. The C++ compiler knows how to optimize the code that contains a temporary variable, i.e. ```a = tmp; a = b; b = tmp;```, but using XOR (might) not. In the end, knowing what is faster is sometimes understanding how a specific compiler/function/library/framework works. what if invoking ```something.length()``` consumes the most amount of time in your ```for``` loop? Why not initialize it beforehand? ```let l = something.complicatedLength(); for(let i = 0; x < l; i++)``` Sure it's an extra line, but it's a line that saves running time!

**Comment by astrocronopio** [comment]
I have seen one-liners of people using std functions to solve katas, it's easy to use a predefined function that already does what you should have coded, it solves the kata, but do you understand the underlying algorithm?  I like writing the first thing that comes to my mind and then trying to optimize :)

**Comment by MarkWilliamsDev** [comment]
I've been coding a few years but never professionally so been wondering the same. Although it's very impressive to see a solution that took me 26 lines of code being done with just one line (not counting first line function declaration and return code), I can't imagine trying to debug or collaborate through code like that. Good to see others also agree on the importance of readability in the real world.

**Comment by umlittlethings** [comment]
yeah
